### PR TITLE
Fixes #26206 - Content view dependency solving

### PR DESCRIPTION
**Pull request by johnpmitsch**
This adds an option to solve RPM dependencies in a content view

- requires latest bastion release, you will have to bundle update

## Creating a content View ##

There is now an option to solve dependencies for the content view when creating one

cv

This is hidden when a composite content view is checked and the auto publish option is shown. Validation is done server-side as well (that a composite CV can't solve dependencies)

cv

You can update this setting in the details too

here

## Global Setting ##

You can set globally to have every content view publish solve dependencies 

set

You can also change from the 'conservative' algorithm to a 'greedy' one in pulp. The default is conservative.

## Skip Dep Solving ##

You can still opt out of the dependency solving at publish time

here

This is hidden until the global or content view setting is turned on.  

I thought about adding an 'opt-in' checkbox when dep solving is not turned on, but I know in the past, users like to check the shiny new checkbox without fully knowing it's trade-offs, so its probably better to keep a more deliberate setting.

## Further things to check ##

- performance impact, how does setting affect performance? This can be used in the messages - "publishes will take approximately 3x as long"
- extra testing with larger repos and more complex filters

### CODE REVIEW

File: app/models/setting/content.rb [hunk 1]
```diff
@@ -9,6 +9,7 @@ def self.load_defaults
 
     download_policies = proc { Hash[::Runcible::Models::YumImporter::DOWNLOAD_POLICIES.map { |p| [p, p] }] }
     proxy_download_policies = proc { Hash[::SmartProxy::DOWNLOAD_POLICIES.map { |p| [p, p] }] }
+    dependency_solving_options = proc { Hash[['conservative', 'greedy'].map { |p| [p, p] }] }
```
Comment (parthaa) [inline app/models/setting/content.rb]: Wonder if this is clearer
```
dependency_solving_options = proc { {'conservative': 'conservative', 'greedy': 'greedy']} }
```

File: app/models/setting/content.rb [hunk 2]
```diff
@@ -101,7 +102,18 @@ def self.load_defaults
...
                  :collection => proc { Hash[Location.unscoped.all.map { |loc| [loc[:title], loc[:title]] }] }),
         self.set('expire_soon_days', N_('The number of days remaining in a subscription before you will be reminded about renewing it.'),
-                 120, N_('Expire soon days'))
+                 120, N_('Expire soon days')),
+        self.set('content_view_solve_dependencies',
+                 N_('Solve RPM dependencies on all Content View publishes. Publishes will take significantly longer with this option turned on ' \
+                 'and filters will be ignored to solve dependencies. Also, certain scenarios can still cause dependency errors. This can be ' \
+                 'set on individual Content Views as well.'),
+                 false, N_('Content View Dependency Solving')),
+        self.set('dependency_solving_logic',
```
Comment (parthaa) [inline app/models/setting/content.rb]: dependency_solving_algorithm

File: app/services/katello/pulp/repository/yum.rb [hunk 3]
```diff
@@ -60,9 +60,24 @@ def regenerate_applicability
           smart_proxy.pulp_api.extensions.repository.regenerate_applicability_by_ids([repo.pulp_id], true)
         end
 
+        def solve_dependencies_config(options, config = {})
```
Comment (parthaa) [inline app/services/katello/pulp/repository/yum.rb]: why does config need to be an extra parameter here. I dont see it being used anywhere

**Comment by parthaa** [comment]
"Copy Content View" not copying the dep solve setting

**Comment by parthaa** [comment]
looks like it will accept a depsolve setting for 'Composite CV'  - Add validation

**Comment by johnpmitsch** [comment]
@parthaa It's still WIP, but I pushed up a commit to address the comments so far!

File: app/models/setting/content.rb (hunk 1 above)
Comment (johnpmitsch) [reply app/models/setting/content.rb]: though still need to try this one

Comment (johnpmitsch) [reply app/models/setting/content.rb]: @parthaa I added a method to handle converting an array to hash with a key/value of each element in the array

**Comment by johnpmitsch** [comment]
I still have to fix some VCR tests and re-record

**Comment by johnpmitsch** [comment]
but other than that, this is ready for review :)

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-views/details/content-view-publish.controller.js [hunk 4]
```diff
@@ -30,17 +35,41 @@ angular.module('Bastion.content-views').controller('ContentViewPublishController
             $scope.working = false;
         }
 
+        $scope.calculateSolveDeps = function(contentViewSolveDep, solveDepSetting, skipSolveDep) {
```
Comment (johnpmitsch) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-views/details/content-view-publish.controller.js]: This isn't used in the view, but I attached it to the `$scope` so I could easily unit test a bunch of scenarios with it. It may not be best practice, but it gives me some good peace of mind knowing this logic is functioning as expected :peace_symbol:

**Comment by sideangleside** [comment]
I dont think I saw this, but is there (or will there be) appropriate Hammer CLI PRs to add these new options to `hammer content-view (create|update)` ?

**Comment by johnpmitsch** [comment]
@sideangleside Hammer will be automagically updated with the new solve dependencies param for CV create, update and CV publish :ballot_box_with_check:

File: app/models/setting/content.rb [hunk 5]
```diff
@@ -101,7 +107,18 @@ def self.load_defaults
                  nil, N_('Default Location Puppet content'), nil,
                  :collection => proc { Hash[Location.unscoped.all.map { |loc| [loc[:title], loc[:title]] }] }),
         self.set('expire_soon_days', N_('The number of days remaining in a subscription before you will be reminded about renewing it.'),
-                 120, N_('Expire soon days'))
+                 120, N_('Expire soon days')),
+        self.set('content_view_solve_dependencies',
+                 N_('Solve RPM dependencies on all Content View publishes. Publishes will take significantly longer with this option turned on ' \
+                 'and filters will be ignored to solve dependencies. Also, certain scenarios can still cause dependency errors. This can be ' \
+                 'set on individual Content Views as well.'),
+                 false, N_('Content View Dependency Solving')),
```
Comment (jlsherrill) [inline app/models/setting/content.rb]: I'm not sure i'm a fan of having this setting for a few reasons:
1) It only works in the UI (api/cli ignores it)
2) Its yet another configuration that duplicates an option on an individual content view
3) I think this is letting the use this when it doesn't make sense, since CVs without filters will still use this (needlessly)

Comment (johnpmitsch) [reply app/models/setting/content.rb]: The suggestion of a global setting came from @sideangleside so I'll let him respond

I think having a better idea of how publish time is affected will help us make this decision too, I'll do some rough benchmarking soon

File: test/services/katello/pulp/repository/yum_test.rb [hunk 6]
```diff
@@ -24,7 +24,7 @@ def setup
         end
 
         def delete_repo(repo)
-          ::ForemanTasks.sync_task(::Actions::Pulp::Repository::Destroy, :pulp_id => repo.pulp_id) rescue ''
+          repo.backend_service(SmartProxy.pulp_master).delete
```
Comment (johnpmitsch) [inline test/services/katello/pulp/repository/yum_test.rb]: @jlsherrill continuing from our discussion on IRC, I had to re-record the tests calling this instead of the task to get the re-record of the tests working. We talk directly to pulp in `create_repo` too so I figure its ok, but let me know if you think there is an issue doing things this way.

Comment (jlsherrill) [reply test/services/katello/pulp/repository/yum_test.rb]: delete repo is an async task, so a race condtion on re-record could cause it to fail. I'd leave it as a sync_task

Comment (johnpmitsch) [reply test/services/katello/pulp/repository/yum_test.rb]: @jlsherrill but thats the thing, it wasn't working as a 'sync_task', I would even check pulp after it had run and succeeded, and the repo was still there. Let me see if I can repro on master

Comment (johnpmitsch) [reply test/services/katello/pulp/repository/yum_test.rb]: On master:

Running with `ktest-live`
```
[vagrant@carrot foreman{develop}]$ bundle exec ./ktest-live ../katello/test/services/katello/pulp/repository/yum_test.rb
/home/vagrant/foreman/lib/foreman.rb:8: warning: already initialized constant Foreman::UUID_REGEXP
/home/vagrant/foreman/lib/foreman.rb:8: warning: previous definition of UUID_REGEXP was here
-- execute("SET CONSTRAINTS ALL DEFERRED;")
   -> 0.0006s
Run options: --seed 54377

# Running:

..E

Error:
Katello::Service::Repository::YumVcrTest#test_create_custom:
RestClient::Conflict: 409 Conflict
    /home/vagrant/katello/app/services/katello/pulp/repository.rb:89:in `create'
    /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:136:in `test_create_custom'
    /home/vagrant/katello/test/support/vcr.rb:24:in `block in run'
    /home/vagrant/katello/test/support/vcr.rb:23:in `run'


bin/rails test /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:125

E

Error:
Katello::Service::Repository::YumVcrCopyTest#test_copy_no_filters:
RestClient::Conflict: 409 Conflict
    /home/vagrant/katello/app/services/katello/pulp/repository.rb:89:in `create'
    /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:35:in `create_repo'
    /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:170:in `setup'
    /home/vagrant/katello/test/support/vcr.rb:24:in `block in run'
    /home/vagrant/katello/test/support/vcr.rb:23:in `run'


bin/rails test /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:180

E

Error:
Katello::Service::Repository::YumVcrCopyTest#test_errata_filter:
RestClient::Conflict: 409 Conflict
    /home/vagrant/katello/app/services/katello/pulp/repository.rb:89:in `create'
    /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:35:in `create_repo'
    /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:170:in `setup'
    /home/vagrant/katello/test/support/vcr.rb:24:in `block in run'
    /home/vagrant/katello/test/support/vcr.rb:23:in `run'


bin/rails test /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:194

E

Error:
Katello::Service::Repository::YumVcrCopyTest#test_copy_rpm_filenames:
RestClient::Conflict: 409 Conflict
    /home/vagrant/katello/app/services/katello/pulp/repository.rb:89:in `create'
    /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:35:in `create_repo'
    /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:170:in `setup'
    /home/vagrant/katello/test/support/vcr.rb:24:in `block in run'
    /home/vagrant/katello/test/support/vcr.rb:23:in `run'


bin/rails test /home/vagrant/katello/test/services/katello/pulp/repository/yum_test.rb:186

.....

Finished in 16.886943s, 0.6514 runs/s, 1.7173 assertions/s.
11 runs, 29 assertions, 0 failures, 4 errors, 0 skips
Coverage report Rcov style generated for Unit Tests to /home/vagrant/foreman/coverage/rcov
Coverage report generated for Unit Tests to /home/vagrant/foreman/coverage. 0.0 / 0.0 LOC (100.0%) covered.
```

Repo isn't being deleted in pulp
```
Mar 12 16:36:57 carrot.jomitsch.example.com pulp[2843]: celery.app.trace:INFO: [76214ed1] Task pulp.server.async.tasks._release_resource[76214ed1-0f62-447a-a115-3a954cac7471] succeeded in 0.00607972500438s: None
Mar 12 16:37:00 carrot.jomitsch.example.com pulp[26779]: pulp.server.webservices.middleware.exception:INFO: Duplicate resource: Fedora_17
Mar 12 16:37:02 carrot.jomitsch.example.com pulp[26779]: pulp.server.webservices.middleware.exception:INFO: Duplicate resource: Fedora_17
Mar 12 16:37:04 carrot.jomitsch.example.com pulp[26779]: pulp.server.webservices.middleware.exception:INFO: Duplicate resource: Fedora_17
Mar 12 16:37:05 carrot.jomitsch.example.com pulp[26779]: pulp.server.webservices.middleware.exception:INFO: Duplicate resource: Fedora_17
```

Comment (johnpmitsch) [reply test/services/katello/pulp/repository/yum_test.rb]: Deleting a repo through the UI seems to work fine, but I was able to repro calling just the task on the rails console too:

repo exists in pulp
```
[vagrant@carrot katello{master}]$ pulp-admin -u admin -p admin repo list | grep zoo3
Display Name:        zoo3
```

delete from pulp in command line
```
irb(main):013:0> Katello::Repository.find(474)
=> #<Katello::Repository id: 474, pulp_id: "74bc51d3-daaf-402b-b98b-98406ca5f588", library_instance_id: nil, content_view_version_id: 8, relative_path: "sync_status_page/Library/custom/custom/zoo3", environment_id: 10, saved_checksum_type: nil, distribution_version: nil, distribution_arch: nil, distribution_bootable: nil, distribution_family: nil, distribution_variant: nil, distribution_uuid: nil, container_repository_name: nil, root_id: 67>
irb(main):014:0> Katello::Repository.find(474).pulp_id
=> "74bc51d3-daaf-402b-b98b-98406ca5f588"
irb(main):015:0> task = ::ForemanTasks.sync_task(::Actions::Pulp::Repository::Destroy, pulp_id: "74bc51d3-daaf-402b-b98b-98406ca5f588")
2019-03-12T17:04:54 [D|app|] Current user set to foreman_admin (admin)
2019-03-12T17:04:54 [D|app|] Current user set to foreman_console_admin (admin)
=> #<ForemanTasks::Task::DynflowTask id: "c6759b19-2682-45d9-99a6-25db7f29cf9c", type: "ForemanTasks::Task::DynflowTask", label: "Actions::Pulp::Repository::Destroy", started_at: "2019-03-12 17:04:54", ended_at: "2019-03-12 17:04:54", state: "stopped", result: "success", external_id: "54848503-ce68-4be6-b7aa-b5cb28bcdb33", parent_task_id: nil, start_at: "2019-03-12 17:04:54", start_before: nil, action: "Destroy">
```

repo is still there

```
[vagrant@carrot katello{master}]$ pulp-admin -u admin -p admin repo list | grep zoo3
Display Name:        zoo3
```

Comment (johnpmitsch) [reply test/services/katello/pulp/repository/yum_test.rb]: I think I got it - the action doesn't take a pulp_id, rather a repository_id

Comment (johnpmitsch) [reply test/services/katello/pulp/repository/yum_test.rb]: that was it! Pushed up a new commit with the change

File: app/models/setting/content.rb (hunk 5 above)
Comment (johnpmitsch) [reply app/models/setting/content.rb]: A few thoughts that are fresh in my brain so I'll share them :)

> I think this is letting the user use dep solving when it doesn't make sense, since CVs without filters will still use this (needlessly)

It would be nice to add an option to not dependency solve if there are no filters (as we discussed on IRC)

> I think this is letting the user use dep solving when it doesn't make sense, since CVs without filters will still use this (needlessly)

This is a fair point, the setting combined with the 'skip dep solving' option does complicate things. I'll think of some better ways to go about this

Comment (sideangleside) [reply app/models/setting/content.rb]: I want to 'idiot proof' the setup as much as possible. If the performance hit is small-ish, I can see users setting this globally (as the ends justify the means). However, if the performance hit is significant, I would be ok not having a global setting.

**Comment by johnpmitsch** [comment]
On an initial test, I'm only seeing a small time increase in publishing a content view, from ~30 seconds (no dep solving) to ~40 seconds (with dep solving)

Let me see if I can get a more conclusive test on a nightly release too.

**Comment by jlsherrill** [comment]
@johnpmitsch make sure you test with a rhel7 or rhel6 rpms repo.  When i did my testing previously i saw about 3x increase in publish time.

**Comment by johnpmitsch** [comment]
@jlsherrill I tested with rhel7 extras and a rhel8, unfortunately I was having mongo issues syncing the larger rpm repos in my dev env. Hopefully the pipeline passes and I can get a nightly box spun up to do more comprehensive testing with larger repos.

**Comment by johnpmitsch** [comment]
@jlsherrill @sideangleside testing with a few large RHEL 7 server repos definitely does show the difference.

This is what I consistently got for publishes of the same CV:
no dep solving: 280-300 seconds (a little under 5 minutes)
with dep solving: 735-755 seconds (about 12 and a half minutes)

So this is around 2.5x the publish time. I do wonder if it grows exponentially (since there wasn't a huge increase in the smaller repos). Maybe QE can do further testing when this gets into builds

Thinking about it, maybe we can get rid of the global setting, it isn't too hard to set the dependency solving on the content view. Though, on the other hand, I could see some users who are willing to bear with the longer publish time wanting to keep that on as a default. Let me know what you think

Either way, I'll think of how to address [the two points made in this comment](https://github.com/Katello/katello/pull/7993#discussion_r265116839).

**Comment by johnpmitsch** [comment]
@sideangleside talking with @jlsherrill about this, we think its best to get rid of the global setting and keep the following:

- Setting dependency solving on the content view
- Being able to opt-out the CV dep solving at publish time.

This will simplify the code a bit and keep things simple for the user too. Let us know what you think since you have a good idea about user's expectations :)

**Comment by johnpmitsch** [comment]
> Also, certain scenarios can still cause dependency errors.

@jlsherrill is there a way to be more specific about this?

I was thinking of mentioning errata if that is what causes the exception `Also, certain scenarios with errata can still cause dependency errors.`

**Comment by johnpmitsch** [comment]
@jlsherrill @parthaa updated to only use the Content View setting and provide a default value for new Content Views  in the settings

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 7]
```diff
@@ -100,6 +100,7 @@ def publish
       task = async_task(::Actions::Katello::ContentView::Publish, @view, params[:description],
                         :major => params[:major],
                         :minor => params[:minor],
+                        :solve_dependencies => @view.solve_dependencies,
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/content_views_controller.rb]: Any reason to pass this in as a param rather than just using the view's value further down in the action?

Comment (johnpmitsch) [reply app/controllers/katello/api/v2/content_views_controller.rb]: I can move it to `katello/app/lib/actions/katello/repository/clone_to_version.rb`, that seems to be the last action the CV is referenced

File: app/services/katello/pulp/repository/yum.rb [hunk 8]
```diff
@@ -75,9 +75,25 @@ def regenerate_applicability
...
+          config
+        end
+
         def copy_contents(destination_repo, options = {})
+          override_config = build_overide_config(options)
           rpm_copy_clauses, rpm_remove_clauses = generate_copy_clauses(options[:filters], options[:rpm_filenames])
-          tasks = [smart_proxy.pulp_api.extensions.rpm.copy(repo.pulp_id, destination_repo.pulp_id, rpm_copy_clauses)]
+          tasks = [smart_proxy.pulp_api.extensions.rpm.copy(repo.pulp_id, destination_repo.pulp_id,
+                   rpm_copy_clauses.merge(:override_config => override_config))]
+
```
Comment (jlsherrill) [inline app/services/katello/pulp/repository/yum.rb]: This method works with CV publish (with filters), but there's another method in repository.rb (copy_units, that takes in AR units, and is used with incrmental update).  Could you adapt that to use conservative as well (if set)?

Comment (jlsherrill) [reply app/services/katello/pulp/repository/yum.rb]: and yes in-retrospect, we probably should have named these methods more differently :)

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 7 above)
Comment (jlsherrill) [reply app/controllers/katello/api/v2/content_views_controller.rb]: sounds good!

**Comment by johnpmitsch** [comment]
@jlsherrill @parthaa I updated to check for the dep solving attribute for a content version to  clone_to_version.rb. I'm not sure I fully get [this comment](https://github.com/Katello/katello/pull/7993#discussion_r267755755)? Incremental update looks like it [uses clone_to_version](https://github.com/Katello/katello/blob/master/app/lib/actions/katello/content_view_version/incremental_update.rb#L70) so it should resolve deps? I can test it out tomorrow.

**Comment by jlsherrill** [comment]
It uses clone_to_version initially for cloning the content view version as it is (an exact copy), but then it uses 'Pulp::Repository::CopyUnits' action to copy specific errata or packages over:

https://github.com/Katello/katello/blob/master/app/lib/actions/katello/content_view_version/incremental_update.rb#L234-L245

**Comment by johnpmitsch** [comment]
@jlsherrill updated!

**Comment by johnpmitsch** [comment]
[test katello]

**Comment by johnpmitsch** [comment]
[test katello]

**Comment by johnpmitsch** [comment]
[test katello]

File: app/lib/actions/pulp/repository/copy_units.rb [hunk 9]
```diff
@@ -19,9 +18,9 @@ def invoke_external_task
 
           units = input[:class_name].constantize.where(:id => input[:unit_ids])
 
-          source_repo.backend_service(SmartProxy.pulp_master).copy_units(target_repo, units,
-                                                                         recursive: input[:recursive],
-                                                                         resolve_dependencies: input[:resolve_dependencies])
+          override_config = Katello::Repository.build_override_config(input)
```
Comment (parthaa) [inline app/lib/actions/pulp/repository/copy_units.rb]: can you make this '::Katello::Repository' instead. My test raised the following error
```2019-04-03T06:04:16 [E|bac|94bdfbab] undefined method `build_override_config' for Actions::Katello::Repository:Module (NoMethodError)
```

**Comment by parthaa** [comment]
It seems like the incremental update via errata is not copying packages over. I am guessing recursive=true has a part. Here are my steps to test this
- Create and sync repo pointing to -> https://partha.fedorapeople.org/test-repos/pteradactyl/
- Create a content view with this repo. Publish and promote to library and dev/other envs. Make a note of the version_id
- update the url in the original repo to -> https://partha.fedorapeople.org/test-repos/pteradactly-with-dino-errata/
- sync the repo again (notice errata + module + packages counts must ve increased by 1)
- run the following command for an incremental update via hammer (change content-view-version-id accordingly.)
```
$ hammer content-view version incremental-update --content-view-version-id=9 --errata-ids='RHEA-2013:0098' --lifecycle-environments='Library,dev' --organization-id=2
```

Note before the refactoring in this PR you should see
```
[....................................................................................................................................................................] [100%]
Content View: koan version 1.1
Added Content:
  Packages:
        pteradactyl-1.2-1.noarch
```
And you should see 23 packages, 7 errata and 15 module streams

With this PR however the output looks like
```
[....................................................................................................................................................................] [100%]
Content View: more version 1.1
Added Content:
```
with no content. In the incremental version you see 22 packages, 7 errata and 15 module streams

Basically the pteradactyl-1.2-1.noarch package did not get copied over. I believe this was a side effect of `recursive: true`. It would copy over the packages that belong to the errata. So we might want to keep it there.,

**Comment by johnpmitsch** [comment]
@parthaa it still should be sending `recursive` to pulp, but now it will be `recursive_conservative` when `conservative` is set. This is from the [override config](https://github.com/Katello/katello/pull/7993/files#diff-2c6c8ef362147f3c5df1b948be20667dR21), which sends [this](https://github.com/Katello/katello/pull/7993/files#diff-f6aa9601a7c1afa89d06151a45d30777R54).

So before incremental update was sending `recursive`, and now it is sending `recursive_conservative`. Do you think what you are seeing is a symptom of that?

**Comment by johnpmitsch** [comment]
@parthaa If you still have the setup handy, could you switch to `greedy` in the settings and retry to see if it pulls in that last package?

**Comment by parthaa** [comment]
recursive greedy or conservative made no difference

**Comment by parthaa** [comment]
here are some call logs after the incremental update command that might be useful ->
Master
```
$ grep recursive /tmp/katello-call-logs 
resource - https://theta.partello.example.com:443, action_name, - /pulp/api/v2/repositories/2-lame-v1_3-24036e9b-27d6-4ea9-804b-b411300be48b/actions/associate/, params - [:post, "{\"source_repo_id\":\"24036e9b-27d6-4ea9-804b-b411300be48b\",\"criteria\":{\"type_ids\":[\"erratum\"],\"filters\":{\"association\":{\"unit_id\":{\"$in\":[\"8daa8118-e180-48f8-8618-f3ce041c058e\"]}}}},\"override_config\":{\"recursive\":true}}", {"content_type"=>"application/json", "accept"=>"application/json"}]
resource - https://theta.partello.example.com:443, action_name, - /pulp/api/v2/repositories/2-lame-v1_3-24036e9b-27d6-4ea9-804b-b411300be48b/actions/associate/, params - [:post, "{\"source_repo_id\":\"24036e9b-27d6-4ea9-804b-b411300be48b\",\"criteria\":{\"type_ids\":[\"modulemd\"],\"filters\":{\"association\":{\"unit_id\":{\"$in\":[\"eab00951-c088-47f4-8fc1-25ebef353e25\"]}}}},\"override_config\":{\"recursive\":true}}", {"content_type"=>"application/json", "accept"=>"application/json"}]
```

With your pr
```
$ grep recursive /tmp/katello-call-logs 
resource - https://theta.partello.example.com:443, action_name, - /pulp/api/v2/repositories/2-lame-v1_4-24036e9b-27d6-4ea9-804b-b411300be48b/actions/associate/, params - [:post, "{\"source_repo_id\":\"2-lame-v1_3-24036e9b-27d6-4ea9-804b-b411300be48b\",\"criteria\":{\"type_ids\":[\"rpm\"],\"filters\":{\"unit\":{\"filename\":{\"$exists\":true}}},\"fields\":{\"unit\":[\"name\",\"epoch\",\"version\",\"release\",\"arch\",\"checksumtype\",\"checksum\"]}},\"override_config\":{\"recursive\":true}}", {"content_type"=>"application/json", "accept"=>"application/json"}]
```

**Comment by johnpmitsch** [comment]
@parthaa So things get a little complicated here, hopefully I can explain this well :)

First off, there was broken logic that needed fixing. I updated this and pushed up in a new commit.

In current master, the [`resolve_dependencies`](https://github.com/Katello/katello/blob/master/app/controllers/katello/api/v2/content_view_versions_controller.rb#L128) param gets passed down the actions until it gets to copy units, where it is [eventually passed to pulp](https://github.com/Katello/katello/blob/master/app/lib/actions/pulp/repository/copy_units.rb#L24). BUT, there is no `resolve_dependencies` param in the pulp API. The params to resolve deps is `recursive` for greedy or `recursive_conservative` for conservative. As you can see, `recursive` is being passed, no matter what the param says. So this means, `resolve_dependencies` param isn't actually doing anything, the incremental update is being performed with greedy dependency resolving no matter what.

So after the updates I just pushed up, the `resolve_dependencies` param now works as expected. The logic is:
```
was resolve_depedencies passed in as a param OR is the content view set to solve dependencies?
if yes - solve dependencies using the algorithm from settings
if no - don't solve dependencies
```

To get the workflow you described working, I had to add `--resolve-dependencies true` param to the hammer command and change the algorithm to `greedy` in the settings.

I think having the algorithm follow settings is fine, but I wonder if we want to always resolve dependencies like we have been doing on incremental update? The way things are in this PR now, we are changing the current behavior, but its changing to make a param actually work.

Let me know if any of this is incoherent, otherwise, let me know what you think!

**Comment by parthaa** [comment]
> I think having the algorithm follow settings is fine, but I wonder if we want to always resolve dependencies like we have been doing on incremental update? The way things are in this PR now, we are changing the current behavior, but its changing to make a param actually work.

I think resolve dependencies should always be true for incremental. Whether we use greedy or conservative is upto the setting. Present behaviour in master reflects that.

**Comment by johnpmitsch** [comment]
@parthaa updated so `resolve-dependencies` API param defaults to true. It still takes into account the setting so only `greedy` will work for your use case.

**Comment by parthaa** [comment]
> @parthaa updated so `resolve-dependencies` API param defaults to true. It still takes into account the setting so only `greedy` will work for your use case.

ok I looked at this. The greedy case works well. I dont like the behaviour of recursive_conservative. It could be just that recursive_conservative does not work for errata in pulp 2.18 and hence I m not seeing the packages or modules (because my dino errata updates an existing package.) 

Will mull over this tomorrow and ack here if its a pulp bug.

**Comment by johnpmitsch** [comment]
@parthaa I'm open to only using `recursive` (greedy) for incremental update esp since we have only been using this already. I'll be curious to hear what the pulp team says though!

**Comment by parthaa** [comment]
[test katello]
